File: tools/python/tests/statistics/summary/test_StreamingEntropy.py
import math
import unittest

from KML.statistics.summary.StreamingEntropy import PyStreamingEntropy


class TestStreamingEntropy(unittest.TestCase):

    WINDOW_SIZE = 10
    NUM_BINS = 10
    SE = None

    def setUp(self):
        self.SE = PyStreamingEntropy(self.WINDOW_SIZE, self.NUM_BINS)

    def tearDown(self):
        self.SE = None

    def test_empty_evaluate(self):
        self.assertEqual(self.SE.evaluate(), 0.0)

    def test_uniform_entropy_with_decrement(self):
        data = list(range(1, 12))
        self.SE.update(data)
        self.assertAlmostEqual(math.log(self.NUM_BINS, 2), self.SE.evaluate())

    def test_uniform_entropy_without_decrement(self):
        data = list(range(1, 6))
        self.SE.update(data)
        self.assertAlmostEqual(math.log(5, 2), self.SE.evaluate())

    def test_uniform_entropy_double_bins(self):
        data = [1, 1, 2, 2, 3, 3, 4, 4, 5, 5]
        self.SE.update(data)
        self.assertAlmostEqual(math.log(5, 2), self.SE.evaluate())

File: tools/cpp/tests/statistics/summary/test_StreamingEntropy.cc
// unittest_streaming_entropy

#include <gtest/gtest.h>
#include "StreamingEntropy.h"
#include <vector>
#include <cmath>

using KML::Statistics::StreamingEntropy;

class StreamingEntropyTest : public ::testing::Test
{
    public:
        StreamingEntropy* se;
        size_t m_windowSize = 10;
        size_t m_numBins = 10;

        StreamingEntropyTest()
        {
            se = new StreamingEntropy(m_windowSize, m_numBins);
        }
};

TEST_F(StreamingEntropyTest, TestEmptyEvaluate)
{
    EXPECT_FLOAT_EQ(0.0, se->evaluate());
}

TEST_F(StreamingEntropyTest, TestUniformEntropyWithDecrement)
{
    std::vector<double> data{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};
    se->update(data);
    EXPECT_FLOAT_EQ(std::log2(m_numBins), se->evaluate());
}

TEST_F(StreamingEntropyTest, TestUniformEntropyWithoutDecrement)
{
    se = new StreamingEntropy(m_windowSize, 5);
    std::vector<double> data = {1, 2, 3, 4, 5};
    se->update(data);
    EXPECT_FLOAT_EQ(std::log2(5), se->evaluate());
}

TEST_F(StreamingEntropyTest, TestUniformEntropyDoubleBins)
{
    // This should be binned with 5 bins.
    std::vector<double> data = {1, 1, 2, 2, 3, 3, 4, 4, 5, 5};
    se->update(data);
    EXPECT_FLOAT_EQ(std::log2(5), se->evaluate());
}

